### ISSUE TITLE
DSA keygen: Sophie-Germain primes not accepted?

### ISSUE THREAD

**Issue by smuellerDD** [usnistgov/ACVP#529]
**testSessionId**
932

**vsId**
4033

**Algorithm registration**
DSA keygen with L=2048 / N=256

When generating DSA keys using Sophie-Germain Primes (MODP 14 to be precise), the ACVP server fails the DSA keygen with the following verdict:

```
}, {
  "vsId" : 4033,
  "disposition" : "failed",
  "tests" : [ {
    "tcId" : 1,
    "result" : "failed",
    "reason" : "Validation failed: Could not find p, q, or g"
  }, {
```

Note, the answer contains the following domain parameters:

```
        "p":"ffffffffffffffff68aaac8a5a8e72151005fa981826d215e56a95ea7c49953918175895f6cb2bdec9524c6ff05dc5b58fa207eca283279b03860e182c779ee33bce362e465e90327c2118ca086c74f10498bc4a4e350c676d9696700729d59ebb52852056f3621c96ada3dc235d65835fcf24fda83f16699ad3551c3648da9805bf63a1b87c00c23d5be4ec51662849e61f4b7c11249faea59f895afb6b38eeedb706f4b65cff0b6bed37a6e9424cf4c67e5e6276b585e445c2516d6d35e14f37145ff26d0a2b301b433acdb31995efdd04348e79084a51229b133ba6be0b0274cc678a084e0229d11cdc808b62c6c434c26821a2da0fc9ffffffffffffffff",
        "q":"0000000000000000000000000000000000000000000000000000000000000000",
        "g":"02000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
```

Note 2: The DSA keygen is performed to suit the dependency for DH. The IUT actually does not implement DSA at all (or a PQG gen for that matter). Thus, to satisfy the dependency, the IUT generates the DH key pair using the MODP14 parameters.

Why would the ACVP server not like that?

**Comment by celic** [comment]
That `q` value only contains 0s, so according to the server a valid value isn't present.

**Comment by smuellerDD** [comment]
Am Freitag, 7. Dezember 2018, 18:33:11 CET schrieb Chris Celi:

Hi Chris,

> That `q` value only contains 0s, so according to the server a valid value
> isn't present.

Well, we look at these Sophie-Germain primes. For example, RFC3526 defining 
these domain parameters, the order Q is never defined. According to our 
understanding it is effectively zero. But we will look into it a bit more.

Note, as SP800-56A rev3 tries to push more and more safe primes, including 
RFC3526, the question around Q will probably be asked by others too.

Ciao
Stephan

**Comment by celic** [comment]
Yeah this sort of 56A rev 3 testing is not implemented yet. Our current DSA tests (especially concerning 800-56A) do not do any work relating to safe primes or Sophie Germain primes. Unfortunately it will be a while before we even approach this as it is considered a "new algorithm".

**Comment by smuellerDD** [comment]
Am Freitag, 7. Dezember 2018, 18:48:39 CET schrieb Chris Celi:

Hi Chris,

> Yeah this sort of 56A rev 3 testing is not implemented yet. Our current DSA
> tests (especially concerning 800-56A) do not do any work relating to safe
> primes or Sophie Germain primes. Unfortunately it will be a while before we
> even approach this as it is considered a "new algorithm".

Hm, for CAVS that is solved somehow - I am not sure how and need to look. But 
the IUT that I am currently looking into does not have any DSA support, but 
offers DH. To satisfy dependencies, DSA keygen must be tested. The IUT does 
not have any means to generate PQG and uses the domain parameters from RFCs 
(MODP14 through 18) to generate DH keys which can be used for covering the DSA 
key gen too. All we need is the ability to report these MODP domain 
parameters.

On the other hand, I do not think that the dependency in my scenario is 
warranted. Thus, if NIST would waive it, we are fine too :-)

Ciao
Stephan

**Comment by celic** [comment]
Can this also be closed via #700?
